crates/dbsp/src/tutorial.rs

```diff
@@ -0,0 +1,744 @@
+//!
+//! ```ignore
+//!         .partitioned_rolling_average(RelRange::new(RelOffset::Before(2), RelOffset::Before(0)))
+//! ```
+//!
+//! We can re-map to recover year and month from the timestamp that we made.  We
+//! also strip off a `Some(...)` that gets wrapped around the rolling average.
+//! This can end up as a `None` if there are no rows in the window, but, in
+//! turn, that can only happen if the window doesn't include the current row,
+//! and we didn't define our window that way.  Anyhow:
+//!
+//! ```ignore
+//!         .map_index(|(c, (date, avg))| ((c.clone(), date / 12, date % 12 + 1), avg.unwrap()));
+//! ```
+//!
+//! We can print the data the same way as before:
+//!
+//! ```ignore
+//! moving_averages.inspect(|data| {
+//!     data.iter()
+//!         .for_each(|((c, y, m), sum, w)| println!("{c:16} {y}-{m:02} {sum:10}: {w:+}"))
+//! });
+//! ```
+//!
+//! The output looks like this (you can verify that the second row is the
+//! average of the first two rows in the previous output, and so on):
+//!
+//! ```text
+//! England          2021-01 5600174: +1
+//! England          2021-02 7488796: +1
+//! England          2021-03 8946255: +1
+//! England          2021-04 10842512: +1
+//! England          2021-05 12307688: +1
+//! England          2021-06 12002268: +1
+//! ...
+//! Northern Ireland 2021-01 150315: +1
+//! Northern Ireland 2021-02 233694: +1
+//! ...
+//! Wales            2021-01 295057: +1
+//! Wales            2021-02 458273: +1
+//! Wales            2021-03 584463: +1
+//! ```
+//!
+//! The whole program is in `tutorial6.rs`.
+//!
+//! ### Joins
+//!
+//! Suppose we want both the current month's vaccination count and the moving
+//! average together. With enough work, we could get them in a linear series of
+//! steps by writing our own "aggregator" ([`Aggregator`]), but it's a little
+//! easier to do a join, and it gives us a chance to show how to do that.  Both
+//! our monthly vaccination counts and our moving averages are indexed Z-sets
+//! with the same key type.
+//!
+//! The first step is to take the code we're written so far and change the final
+//! [`map_index`](`FilterMap::map_index`) on `moving_averages` to include a
+//! couple of casts, so that `monthly_totals` and `moving_averages` have exactly
+//! the same key type (both `(String, i32, u8)`).  The new call looks like this;
+//! only the `as <type>` parts are new:
+//!
+//! ```ignore
+//!         .map_index(|(c, (date, avg))| {
+//!             (
+//!                 (c.clone(), (date / 12) as i32, (date % 12 + 1) as u8),
+//!                 avg.unwrap(),
+//!             )
+//!         });
+//! ```
+//!
+//! Then we can use [`join_index`](`Stream::join_index`) to do the join and
+//! [`inspect`](`Stream::inspect`) to print the results, as shown below.  The
+//! only subtlety is the use of `Some(...)`, which is here because the closure
+//! passed to `join_index` must return an iterable collection and `Some(x)` is a
```

Comment:
Maybe it's convenient, but for someone who doesn't know Rust it can be pretty confusing.
Why not just remove the Some()?
I think the join signature also requires a little explanation.